### ISSUE TITLE
Anomaly Detection

### ISSUE THREAD

**Issue by WatershedXiaolan** [WatershedXiaolan/Xiaolan-s-Gitbook#86]
### Univariate Anomaly Detection
mean, std, skew, kurt, scatter, Isolation forest

### Multivariate Anomaly Detection
KNN: its distance to its kth nearest neighbor could be viewed as the outlier score.
Probability distribution 

### ANN: An autoencoder is a type of artificial neural network used to learn efficient data codings in an unsupervised manner. The aim of an autoencoder is to learn a representation (encoding) for a set of data, typically for dimensionality reduction. Along with the reduction side, a reconstructing side is learnt, where the autoencoder tries to generate from the reduced encoding a representation as close as possible to its original input.

![](https://miro.medium.com/max/1400/1*QjevZB4Qg5Su7ufVfZ4Rwg.jpeg)

In the context of anomaly detection and condition monitoring, the basic idea is to use the autoencoder network to “compress” the sensor readings to a lower-dimensional representation, which captures the correlations and interactions between the various variables. (Essentially the same principle as the PCA model, but here we also allow for non-linear interactions between the variables).
By monitoring the re-construction error, one can thus get an indication of the “health” of the monitored equipment, as this error will increase as the equipment degrades.

**Comment by WatershedXiaolan** [comment]
Z-Score pros:
It is a very effective method if you can describe the values in the feature space with a gaussian distribution. (Parametric)
The implementation is very easy using pandas and scipy.stats libraries.
Z-Score cons:
It is only convenient to use in a low dimensional feature space, in a small to medium sized dataset.
Is not recommended when distributions can not be assumed to be parametric.

**Comment by WatershedXiaolan** [comment]
Dbscan pros:
It is a super effective method when the distribution of values in the feature space can not be assumed.
Works well if the feature space for searching outliers is multidimensional (ie. 3 or more dimensions)
Sci-kit learn’s implementation is easy to use and the documentation is superb.
Visualizing the results is easy and the method itself is very intuitive.

Dbscan cons:
The values in the feature space need to be scaled accordingly.
Selecting the optimal parameters eps, MinPts and metric can be difficult since it is very sensitive to any of the three params.
It is an unsupervised model and needs to be re-calibrated each time a new batch of data is analyzed.
It can predict once calibrated but is strongly not recommended.

**Comment by WatershedXiaolan** [comment]
Isolation Forest pros:
There is no need of scaling the values in the feature space.
It is an effective method when value distributions can not be assumed.
It has few parameters, this makes this method fairly robust and easy to optimize.
Scikit-Learn’s implementation is easy to use and the documentation is superb.
Isolation Forest cons:
The Python implementation exists only in the development version of Sklearn.
Visualizing results is complicated.
If not correctly optimized, training time can be very long and computationally expensive.
